src/lib.rs

```diff
@@ -490,7 +490,78 @@ impl Qbit {
     }
 
     pub async fn add_torrent(&self, arg: impl Borrow<AddTorrentArg> + Send + Sync) -> Result<()> {
-        self.post("torrents/add", Some(arg.borrow())).await?.end()
+        let a: &AddTorrentArg = arg.borrow();
+        match &a.source {
+            TorrentSource::Urls { urls: _ } => {
+                self.post("torrents/add", Some(arg.borrow())).await?.end()
+            }
+            TorrentSource::TorrentFiles { torrents } => {
+                {
+                    for i in 0..3 {
+                        // If it's not the first attempt, we need to re-login
+                        self.login(i != 0).await?;
+                        let mut form = reqwest::multipart::Form::new();
+                        // Serialize `a` to a HashMap and add it to the form
+                        let a_map: HashMap<String, String> = {
```

Comment:
I prefer handwriting this part, and not all fields are string

Reply:
Considering that the `Form::text` method actually accepts `str`
```Rust
reqwest::multipart::Form::text<T, U>(self, name: T, value: U) -> Form
    where
        T: Into<Cow<'static, str>>,
        U: Into<Cow<'static, str>>,
```
Maybe we can be lazy with `serde_json`.I think it's the same as `req.form(body)`(except it's `serde_urlencoded` there).
It shouldn't be possible for the serialization to go wrong here, so I removed the extra expect.
```Rust
let a_map: HashMap<String, Value> = serde_json::from_value(serde_json::to_value(&a)?)?;
    for (k, v) in a_map {
        form = form.text(k, v.to_string());
}
```

Reply:
What about keep the as_object, but use for_each and to_string to push the options?

Reply:
there's some ownership troubles here.`Form::text`require the ownership of Form.

Reply:
`form = form.text(...)` then

Reply:
Like this:
```rust
let form = torrents.iter().fold(
    serde_json::to_value(a)?
        .as_object()
        .unwrap()
        .into_iter()
        .fold(reqwest::multipart::Form::new(), |form, (k, v)| {
            form.text(k.to_owned(), v.to_string())
        }),
    |mut form, torrent| {
        let p = reqwest::multipart::Part::bytes(torrent.data.clone())
            .file_name(torrent.filename.clone())
            .mime_str("application/x-bittorrent")
            .unwrap();
        form = form.part("torrents", p);
        form
    },
);
```

Reply:
This looks good to me